cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java

```diff
@@ -197,23 +201,63 @@ public void setStatus(String status) {
                                      DiscoveryServiceClient discoveryServiceClient, QueueAdmin queueAdmin,
                                      Scheduler scheduler, PreferencesStore preferencesStore,
-                                     NamespacedLocationFactory namespacedLocationFactory) {
+                                     NamespacedLocationFactory namespacedLocationFactory, MRJobClient mrJobClient) {
     super(authenticator);
     this.namespacedLocationFactory = namespacedLocationFactory;
     this.store = storeFactory.create();
     this.workflowClient = workflowClient;
-    this.configuration = configuration;
+    this.hConf = hConf;
+    this.cConf = cConf;
     this.runtimeService = runtimeService;
-    this.appFabricDir = this.configuration.get(Constants.AppFabric.OUTPUT_DIR);
+    this.appFabricDir = this.cConf.get(Constants.AppFabric.OUTPUT_DIR);
     this.discoveryServiceClient = discoveryServiceClient;
     this.queueAdmin = queueAdmin;
     this.scheduler = scheduler;
     this.preferencesStore = preferencesStore;
-    this.schedulerQueueResolver = new SchedulerQueueResolver(configuration, store);
+    this.schedulerQueueResolver = new SchedulerQueueResolver(cConf, store);
+    this.mrJobClient = mrJobClient;
+  }
+
+  /**
+   * Relays job-level and task-level information about a particular MapReduce program run.
+   */
+  @GET
+  @Path("/apps/{app-id}/mapreduces/{mapreduce-id}/runs/{run-id}/info")
+  public void mapReduceInfo(HttpRequest request, HttpResponder responder,
+                            @PathParam("namespace-id") String namespaceId,
+                            @PathParam("app-id") String appId,
+                            @PathParam("mapreduce-id") String mapreduceId,
+                            @PathParam("run-id") String runId) {
+    try {
+      Id.Program programId = Id.Program.from(namespaceId, appId, ProgramType.MAPREDUCE, mapreduceId);
+      ApplicationSpecification appSpec = store.getApplication(programId.getApplication());
+      if (appSpec == null) {
+        responder.sendString(HttpResponseStatus.NOT_FOUND, String.format("Application not found: %s",
+                                                                         programId.getApplication()));
+        return;
+      }
+      if (!appSpec.getMapReduce().containsKey(mapreduceId)) {
+        responder.sendString(HttpResponseStatus.NOT_FOUND, String.format("Program not found: %s", programId));
+        return;
+      }
+
+      MRJobInfo mrJobInfo = mrJobClient.getMRJobInfo(runId);
+      responder.sendJson(HttpResponseStatus.OK, mrJobInfo);
+    } catch (NotFoundException e) {
+      LOG.debug("RunId not found: {}", runId, e);
+      responder.sendString(HttpResponseStatus.NOT_FOUND, e.getMessage());
+    } catch (IOException ioe) {
+      LOG.warn("Failed to get run history for runId: {}", runId, ioe);
+      responder.sendStatus(HttpResponseStatus.SERVICE_UNAVAILABLE);
```

Comment:
IOE to not signify internal server error, but that some service was not fully available to communicate with.